Rewrite REST examples to use QS + Misc fixes (#541)
strapi/documentation: docs/developer-docs/latest/developer-resources/database-apis-reference/rest/filtering-locale-publication.md

---
title: Filtering for REST API - Strapi Developer Docs
description: Use Strapi's REST API to filter the results of your requests.
sidebarDepth: 3
canonicalUrl: https://docs.strapi.io/developer-docs/latest/developer-resources/database-apis-reference/rest-api/filtering-locale-publication.html
---

# REST API: Filtering, Locale, and Publication State

> The [REST API](/developer-docs/latest/developer-resources/database-apis-reference/rest-api.md) offers the ability to filter results found with its ["Get entries"](/developer-docs/latest/developer-resources/database-apis-reference/rest-api.md#get-entries) method. If the [Internationalization (i18n) plugin](/developer-docs/latest/plugins/i18n.md) is enabled on your content-type it's possible to filter by locale and likewise if you have [Draft & Publish](/developer-docs/latest/concepts/draft-and-publish.md) enabled it's possible to filter based on a `live` or `draft` state.

Reviewer: ```suggestion
The [REST API](/developer-docs/latest/developer-resources/database-apis-reference/rest-api.md) offers the ability to filter results found with its ["Get entries"](/developer-docs/latest/developer-resources/database-apis-reference/rest-api.md#get-entries) method.<br/>
Using optional Strapi features can provide some more filters:

* If the [Internationalization (i18n) plugin](/developer-docs/latest/plugins/i18n.md) is enabled on a content-type, it's possible to filter by locale.
* If the [Draft & Publish](/developer-docs/latest/concepts/draft-and-publish.md)is  enabled, it's possible to filter based on a `live` or `draft` state.
```

Cleaner look (using lists) when we have multiple possibilities + avoid using "you" (neutral tone, see our Handbook > 12 rules)

```json
[
  {
    "id": 1,
    "username": "John",
    "email": "john@test.com",
    "provider": "local",
    "confirmed": true,
    "blocked": false,
    "createdAt": "2021-12-03T20:08:17.740Z",
    "updatedAt": "2021-12-03T20:08:17.740Z"
  }
]
```

:::
::::

::::api-call
:::request Example request: Find multiple restaurants with ids 3, 6, 8

```js
const qs = require('qs');
const query = qs.stringify({
  filters: {
    id: {
      $in: [3, 6, 8],
    },
  },
}, {
  encodeValuesOnly: true, // prettify url
});

await request(`/api/restaurants?${query}`);
// GET /api/restaurants?filters[id][$in][0]=3&filters[id][$in][1]=6&filters[id][$in][2]=8
```

:::

:::response Example response

```json
{
  "data": [
    {
      "id": 3,
      "attributes": {
        "name": "test3",
        // ...
      }
    },
    {
      "id": 6,
      "attributes": {
        "name": "test6",
        // ...
      }
    },
    {
      "id": 8,
      "attributes": {
        "name": "test8",
        // ...
      }
    }
  ],
  "meta": {
    // ...
  }
}
```

:::
::::

:::caution
By default, the filters can only be used from `find` endpoints generated by the Content-Type Builder and the CLI.
:::

> ### Complex Filtering

Reviewer: ```suggestion
### Complex filtering
```

Styleguide - We don't title case every word in h2/h3/h4 headings

> Complex filtering is combining multiple filters utilizing advanced methods such as combining `$and` & `$or`. This allows for an extreme amount of flexibility to request exactly the data needed.

Reviewer: ```suggestion
Complex filtering is combining multiple filters utilizing advanced methods such as combining `$and` & `$or`. This allows for more flexibility to request exactly the data needed.
```

Tone of voice - we try to stay neutral

:::
::::

### Deep filtering

Deep filtering is filtering on a relation's fields.

::::api-call
:::request Example request: Find restaurants owned by a chef who belongs to a 5-star restaurant

```js
const qs = require('qs');
const query = qs.stringify({
  filters: {
    chef: {
      restaurants: {
        stars: {
          $eq: 5,
        },
      },
    },
  },
}, {
  encodeValuesOnly: true,
});

await request(`/api/restaurants?${query}`);
// GET /api/restaurants?filters[chef][restaurants][stars][$eq]=5
```

:::

:::response Example response

```json
{
  "data": [
    {
      "id": 1,
      "attributes": {
        "name": "GORDON RAMSAY STEAK",
        "stars": 5
        // ...
      }
    },
    {
      "id": 2,
      "attributes": {
        "name": "GORDON RAMSAY BURGER",
        "stars": 5
        // ...
      }
    }
  ],
  "meta": {
    // ...
  }
}
```

:::
::::

::: caution

- Querying your API with deep filters may cause performance issues.  If one of your deep filtering queries is too slow, we recommend building a custom route with an optimized version of the query.
- Deep filtering isn't available for polymorphic relations (eg: Dynamic Zones & Media Fields).

:::

> :::tip

Reviewer: ```suggestion
:::note
```

> Relations, Media Fields, Components, and Dynamic Zones are not populated by default, to populate these please see our [population documentation](/developer-docs/latest/developer-resources/database-apis-reference/rest/populating-fields.md#population). Also please note that it is not possible to filter on Dynamic Zones or Media Fields.

Reviewer: ```suggestion
* Relations, media fields, components, and dynamic zones are not populated by default. Use the `populate` parameter to populate these data structures (see [population documentation](/developer-docs/latest/developer-resources/database-apis-reference/rest/populating-fields.md#population))

* It is not possible to filter on dynamic zones or media fields.
```

2 distinct ideas + casing + formatting/consistency with other links (see [handbook](https://handbook.strapi.io/user-success-manual/strapi-documentation-style-guide#7d0e888be1964efb96428ba804be4003) :-) )
